### ISSUE TITLE
TypeError: Cannot read properties of null (reading 'constructor')

### ISSUE THREAD

**Issue by c100k** [remojansen/InversifyJS#1664]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current behavior

When executing tests via vitest, I have the following failure (that is not present in 6.0.3):

```typescript
TypeError: Cannot read properties of null (reading 'constructor')
 ❯ getBaseClassDependencyCount node_modules/src/planning/reflection_utils.ts:22:32
 ❯ forEach node_modules/src/planning/planner.ts:268:50
 ❯ _createSubRequests node_modules/src/planning/planner.ts:239:18
 ❯ forEach node_modules/src/planning/planner.ts:282:9
 ❯ forEach node_modules/src/planning/planner.ts:281:20
 ❯ _createSubRequests node_modules/src/planning/planner.ts:239:18
 ❯ node_modules/src/planning/planner.ts:282:9
```

### Expected behavior

Ideally, the code shouldn't fail like this. Indeed, there is no problem with `6.0.3`. From `6.1.0` and subsequent versions it fails. These changes introduced in `6.1.0` in https://github.com/inversify/InversifyJS/blob/master/src/planning/reflection_utils.ts#L20 seem very hacky with the `as` and the eslint rules disabled.

On the other side, when the error occurs, the stacktrace should give the correct paths. Right now, it gives `node_modules/src/planning/reflection_utils.ts` which is not the correct path. I think there is a problem with the sourcemap generation which goes one directory too far because of `../../../`.

Additionnally, I'm not sure it's a good idea to ship `esm` already bundled. Indeed, it makes it almost impossible to debug internally in order to find out the underlying issue of such an error.

### Possible solution

I suspect there is an issue with constructor arguments that are not "injected". But I'm still digging to understand why it works on `6.0.3` and why it has stopped working with `6.1.0` and subsequent versions.

When I add `console.log` in the bundled ESM code, I'm getting this:

```typescript
{ metadataReader: E {}, func: [class Getter], t: 'function' }
{ metadataReader: E {}, func: [Function: Object], t: 'function' }
```

Indeed, this function seem to be able to receive something else than a `Newable` so it must make additional checks and avoid at all prices "casting" the type.

### Package version

6.1.5

### Node.js version

22

### In which operating systems have you tested?

- [X] macOS
- [ ] Windows
- [ ] Linux

**Comment by notaphplover** [comment]
Hey @c100k, thank you for your report. Please submit a minimum recreation code via code sandbox or any other platform. A reproduction repository with the minimum reproduction code is a valid choice as well.

**Comment by notaphplover** [comment]
I see the issue, on my way to patch it.

> Additionnally, I'm not sure it's a good idea to ship esm already bundled. Indeed, it makes it almost impossible to debug internally in order to find out the underlying issue of such an error.

I wish there was a better way to provide an ESM build compatible with NodeJS, browsers and bundlers in any typescript config. Not my cup of tea as well, but I don't see a better approach.

> On the other side, when the error occurs, the stacktrace should give the correct paths. Right now, it gives node_modules/src/planning/reflection_utils.ts which is not the correct path. I think there is a problem with the sourcemap generation which goes one directory too far because of ../../../.

I see, indeed, `esm` source maps have the wrong parent directory :thinking:, it seems rollup is not properly generating them. I'll patch that as well

**Comment by notaphplover** [comment]
Fixed in #1665, released in `inversify@6.1.6`. Closing the issue now. If the issue persist, just submit a recreation code and I'll reopen it.
